server/common/immutable_kvcache.py

```diff
@@ -11,14 +11,44 @@ class ImmutableKVCache(MutableMapping):
     def __init__(self, factory):
         self.factory = factory
         self.lock = threading.Lock()
+        self.factory_calls = {}
         self.cache = {}
         super().__init__()
 
     def __getitem__(self, key):
-        if key not in self.cache:
-            with self.lock:
-                if key not in self.cache:
+        if key in self.cache:
+            return self.cache[key]
+
+        # we need to call factory.  First grab the main lock and the per-key CV.
+        factory_calls = None
+        with self.lock:
+            if key in self.cache:
+                return self.cache[key]
+            if key not in self.factory_calls:
+                creation_thr = True
+                self.factory_calls[key] = {'cv': threading.Condition(), 'is_done': False, 'error': None}
+            factory_calls = self.factory_calls[key]
+
+        # with the CV, create the value (or wait for it to be created)
+        cv = factory_calls['cv']
+        with cv:
+            if creation_thr:
+                try:
                     self.cache[key] = self.factory(key)
+                except Exception as e:
+                    factory_calls['error'] = e
+
+                factory_calls['is_done'] = True
+                cv.notify_all()
+            else:
+                """ wait for the value to be available """
+                while not factory_calls['is_done']:
+                    cv.wait()
+
+        with self.lock:
+            if key in self.factory_calls:
+                del self.factory_calls[key]
+
         return self.cache[key]
```

Comment:
This could get a KeyError if __delitem__ was called between when this was set and when it was accessed here.
Maybe access to self.cache should have a lock.

Reply:
as the cxg adaptor uses it, it won't ever have this issue (it is only called from CxgAdaptor.clean()).  But to make this general-purpose, you have a great point.